prusti-common/src/vir/vir_macro.rs

```diff
@@ -0,0 +1,125 @@
+            name: stringify!($name).to_string(),
+            typ: vir_type!($type)
+        }
+    }
+}
+
+
+macro_rules! vir {
+    (# $comment: expr) => {
+        crate::vir::Stmt::Comment($comment.to_string())
+    };
+    (label $label: expr) => {
+        crate::vir::Stmt::Label($label.to_string())
+    };
+    (assert $exp: tt) => {
+        crate::vir::Stmt::Assert(
+            vir!($exp),
+            crate::vir::FoldingBehaviour::Expr,
+            crate::vir::Position::default())
+    };
+    (inhale $exp: tt) => {
+        crate::vir::Stmt::Inhale(
+            vir!($exp),
+            crate::vir::FoldingBehaviour::Expr)
+    };
+    (apply $exp: tt) => {
+        crate::vir::Stmt::ApplyMagicWand(
+            vir!($exp),
+            crate::vir::Position::default())
+    };
+    (obtain $exp: tt) => {
+        crate::vir::Stmt::Obtain(
+            vir!($exp),
+            crate::vir::Position::default())
+    };
+    (if ($exp: tt) { $($then: tt);* } else { $($elze: tt);* }) => {
+        crate::vir::Stmt::If(vir!($exp), vec![$(vir!($then)),*], vec![$(vir!($elze)),*])
+    };
+    (if ($exp: tt) { $($then: tt);* }) => {
+        crate::vir::Stmt::If(vir!($exp), vec![$(vir!($then)),*], vec![])
+    };
+    (true) => {
+        crate::vir::Expr::Const(
+            crate::vir::Const::Bool(true),
+            crate::vir::Position::default())
+    };
+    (false) => {
+        crate::vir::Expr::Const(
+            crate::vir::Const::Bool(false),
+            crate::vir::Position::default())
+    };
+    ($lhs: tt == $rhs: tt) => {
+        crate::vir::Expr::BinOp(vir::BinOpKind::Equals,
+            Box::new(vir!($lhs)),
+            Box::new(vir!($rhs)),
+            crate::vir::Position::default())
+    };
+    ($head: tt && $tail: tt) => {
+        crate::vir::Expr::BinOp(vir::BinOpKind::And,
+            Box::new(vir!($head)),
+            Box::new(vir!($tail)),
+            crate::vir::Position::default())
+    };
+    ($head: tt || $tail: tt) => {
+        crate::vir::Expr::BinOp(vir::BinOpKind::Or,
+            Box::new(vir!($head)),
+            Box::new(vir!($tail)),
+            crate::vir::Position::default())
+    };
+    ($antecedent: tt ==> $consequent: tt) => {
+        crate::vir::Expr::BinOp(vir::BinOpKind::Implies,
+            Box::new(vir!($antecedent)),
+            Box::new(vir!($consequent)),
+            crate::vir::Position::default())
+    };
+    ($lhs: tt {$borrow: expr} --* $rhs: tt) => {
+        crate::vir::Expr::magic_wand(vir!($lhs), vir!($rhs), $borrow)
+    };
+
+    (forall $($name: ident : $type: ident),+ :: {$trigger: tt} $body: tt) => {
+        crate::vir::Expr::ForAll(
+            vec![$(vir_local!($name: $type)),+],
+            vec![crate::vir::Trigger::new(vec![vir!($trigger)])],
+            Box::new(vir!($body)),
+            crate::vir::Position::default())
+    };
+    ([ $e: expr ]) => { $e.clone() };
+    (( $($tokens: tt)+ )) => { vir!($($tokens)+) }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Please give this a more descriptive name; when this fails when running all tests in Prusti, it just says "failed running tests" or something similar.

Reply:
What's the error message? I think that the convention is to always use `mod test { .. }`, because having unique names for each test module sounds difficult to enforce.

Reply:
@fpoli i can remove the `s` if you want
@Aurel300  the error seems pretty clear to me; it includes the full module path (except the crate name, but that's not something i can fix with the module name here):

```
failures:
    vir::vir_macro::tests::forall

test result: FAILED. 9 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.04s
```

Reply:
Ok, this is fine then. I remember dealing with failures that just said `tests` instead of the full module path; I guess this happened when the `tests` module was top-level in its crate, since the crate name is not shown.